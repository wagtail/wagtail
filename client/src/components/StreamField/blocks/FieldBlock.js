/* global $ */

import { escapeHtml as h } from '../../../utils/text';

export class FieldBlock {
  constructor(blockDef, placeholder, prefix, initialState, initialError) {
    this.blockDef = blockDef;
    this.type = blockDef.name;

    const dom = $(`
      <div class="${h(this.blockDef.meta.classname)}">
        <div class="field-content">
          <div class="input">
            <div data-streamfield-widget></div>
            <span></span>
          </div>
        </div>
      </div>
    `);
    $(placeholder).replaceWith(dom);
    const widgetElement = dom.find('[data-streamfield-widget]').get(0);
    this.element = dom[0];

    try {
      this.widget = this.blockDef.widget.render(
        widgetElement,
        prefix,
        prefix,
        initialState,
      );
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error(e);
      this.setError([
        {
          messages: [
            'This widget failed to render, please check the console for details',
          ],
        },
      ]);
      return;
    }

    this.idForLabel = this.widget.idForLabel;

    if (this.blockDef.meta.helpText) {
      const helpElement = document.createElement('p');
      helpElement.classList.add('help');
      helpElement.innerHTML = this.blockDef.meta.helpText; // unescaped, as per Django conventions
      this.element.querySelector('.field-content').appendChild(helpElement);
    }

    if (window.comments && this.blockDef.meta.showAddCommentButton) {
      const fieldCommentControlElement = document.createElement('div');
      fieldCommentControlElement.classList.add('field-comment-control');
      this.element.appendChild(fieldCommentControlElement);

      const addCommentButtonElement = document.createElement('button');
      addCommentButtonElement.type = 'button';
      addCommentButtonElement.setAttribute(
        'aria-label',
        blockDef.meta.strings.ADD_COMMENT,
      );
      addCommentButtonElement.setAttribute('data-comment-add', '');
      addCommentButtonElement.classList.add('button');
      addCommentButtonElement.classList.add('button-secondary');
      addCommentButtonElement.classList.add('button-small');
      addCommentButtonElement.classList.add('u-hidden');
      addCommentButtonElement.innerHTML =
        '<svg class="icon icon-comment-add initial icon-default" aria-hidden="true" focusable="false">' +
        '<use href="#icon-comment-add"></use></svg>' +
        '<svg class="icon icon-comment-add initial icon-reversed" aria-hidden="true" focusable="false">' +
        '<use href="#icon-comment-add-reversed"></use></svg>';
      fieldCommentControlElement.appendChild(addCommentButtonElement);
      window.comments.initAddCommentButton(addCommentButtonElement);
    }

    if (initialError) {
      this.setError(initialError);
    }
  }

  setState(state) {
    if (this.widget) {
      this.widget.setState(state);
    }
  }

  setError(errorList) {
    this.element
      .querySelectorAll(':scope > .field-content > .error-message')
      .forEach((element) => element.remove());

    if (errorList) {
      this.element.classList.add('error');

      const errorElement = document.createElement('p');
      errorElement.classList.add('error-message');
      errorElement.innerHTML = errorList
        .map((error) => `<span>${h(error.messages[0])}</span>`)
        .join('');
      this.element.querySelector('.field-content').appendChild(errorElement);
    } else {
      this.element.classList.remove('error');
    }
  }

  getState() {
    return this.widget.getState();
  }

  getValue() {
    return this.widget.getValue();
  }

  getTextLabel(opts) {
    if (this.widget.getTextLabel) {
      return this.widget.getTextLabel(opts);
    }
    return null;
  }

  focus(opts) {
    if (this.widget) {
      this.widget.focus(opts);
    }
  }
}

export class FieldBlockDefinition {
  constructor(name, widget, meta) {
    this.name = name;
    this.widget = widget;
    this.meta = meta;
  }

  render(placeholder, prefix, initialState, initialError) {
    return new FieldBlock(
      this,
      placeholder,
      prefix,
      initialState,
      initialError,
    );
  }
}
