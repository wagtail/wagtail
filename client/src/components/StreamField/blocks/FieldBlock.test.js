/* eslint-disable @typescript-eslint/no-unused-vars */

import { FieldBlockDefinition } from './FieldBlock';

import $ from 'jquery';
window.$ = $;

window.comments = {
  getContentPath: jest.fn(),
};

// Define some callbacks in global scope that can be mocked in tests
let constructor = (_widgetName, _name, _id, _initialState) => {};
let setState = (_widgetName, _state) => {};
let getState = (_widgetName) => {};
let getValue = (_widgetName) => {};
let focus = (_widgetName) => {};

class DummyWidgetDefinition {
  constructor(widgetName, { throwErrorOnRender = false } = {}) {
    this.widgetName = widgetName;
    this.throwErrorOnRender = throwErrorOnRender;
  }

  render(placeholder, name, id, initialState) {
    if (this.throwErrorOnRender) {
      throw new Error();
    }

    const widgetName = this.widgetName;
    constructor(widgetName, { name, id, initialState });

    $(placeholder).replaceWith(`<p name="${name}" id="${id}">${widgetName}</p>`);
    return {
      setState(state) { setState(widgetName, state); },
      getState() { getState(widgetName); return `state: ${widgetName} - ${name}`; },
      getValue() { getValue(widgetName); return `value: ${widgetName} - ${name}`; },
      focus() { focus(widgetName); },
      idForLabel: id,
    };
  }
}

class ValidationError {
  constructor(messages) {
    this.messages = messages;
  }
}

describe('telepath: wagtail.blocks.FieldBlock', () => {
  let boundBlock;

  window.comments = {
    initAddCommentButton: jest.fn(),
    getContentPath: jest.fn(),
  };

  beforeEach(() => {
    // Create mocks for callbacks
    constructor = jest.fn();
    setState = jest.fn();
    getState = jest.fn();
    getValue = jest.fn();
    focus = jest.fn();

    // Define a test block
    const blockDef = new FieldBlockDefinition(
      'test_field',
      new DummyWidgetDefinition('The widget'),
      {
        label: 'Test Field',
        required: true,
        icon: 'placeholder',
        classname: 'field char_field widget-text_input fieldname-test_charblock',
        helpText: 'drink <em>more</em> water'
      }
    );

    // Render it
    document.body.innerHTML = '<div id="placeholder"></div>';
    boundBlock = blockDef.render(
      $('#placeholder'),
      'the-prefix',
      'Test initial state'
    );
  });

  test('it renders correctly', () => {
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  test('constructor is called with correct parameters', () => {
    expect(constructor.mock.calls.length).toBe(1);
    expect(constructor.mock.calls[0][0]).toBe('The widget');
    expect(constructor.mock.calls[0][1]).toEqual({
      name: 'the-prefix',
      id: 'the-prefix',
      initialState: 'Test initial state',
    });
  });

  test('getValue() calls widget getValue()', () => {
    const value = boundBlock.getValue();
    expect(getValue.mock.calls.length).toBe(1);
    expect(value).toEqual('value: The widget - the-prefix');
  });

  test('getState() calls widget getState()', () => {
    const state = boundBlock.getState();
    expect(getState.mock.calls.length).toBe(1);
    expect(state).toEqual('state: The widget - the-prefix');
  });

  test('setState() calls widget setState()', () => {
    boundBlock.setState('Test changed state');
    expect(setState.mock.calls.length).toBe(1);
    expect(setState.mock.calls[0][0]).toBe('The widget');
    expect(setState.mock.calls[0][1]).toBe('Test changed state');
  });

  test('focus() calls widget focus()', () => {
    boundBlock.focus();
    expect(focus.mock.calls.length).toBe(1);
    expect(focus.mock.calls[0][0]).toBe('The widget');
  });

  test('setError() renders errors', () => {
    boundBlock.setError([
      new ValidationError(['Field must not contain the letter E']),
      new ValidationError(['Field must contain a story about kittens']),
    ]);
    expect(document.body.innerHTML).toMatchSnapshot();
  });
});

describe('telepath: wagtail.blocks.FieldBlock with comments enabled', () => {
  let boundBlock;

  window.comments = {
    initAddCommentButton: jest.fn(),
    getContentPath: jest.fn(),
  };

  beforeEach(() => {
    // Create mocks for callbacks
    constructor = jest.fn();
    setState = jest.fn();
    getState = jest.fn();
    getValue = jest.fn();
    focus = jest.fn();

    // Define a test block
    const blockDef = new FieldBlockDefinition(
      'test_field',
      new DummyWidgetDefinition('The widget'),
      {
        label: 'Test Field',
        required: true,
        icon: 'placeholder',
        classname: 'field char_field widget-text_input fieldname-test_charblock',
        helpText: 'drink <em>more</em> water',
        showAddCommentButton: true,
        strings: {
          ADD_COMMENT: 'Add Comment'
        }
      }
    );

    // Render it
    document.body.innerHTML = '<div id="placeholder"></div>';
    boundBlock = blockDef.render(
      $('#placeholder'),
      'the-prefix',
      'Test initial state'
    );
  });

  test('it renders correctly', () => {
    expect(document.body.innerHTML).toMatchSnapshot();
  });
});

describe('telepath: wagtail.blocks.FieldBlock catches widget render errors', () => {
  let boundBlock;

  beforeEach(() => {
    // Create mocks for callbacks
    constructor = jest.fn();
    setState = jest.fn();
    getState = jest.fn();
    getValue = jest.fn();
    focus = jest.fn();

    // Define a test block
    const blockDef = new FieldBlockDefinition(
      'test_field',
      new DummyWidgetDefinition('The widget', { throwErrorOnRender: true }),
      {
        label: 'Test Field',
        required: true,
        icon: 'placeholder',
        classname: 'field char_field widget-text_input fieldname-test_charblock',
        helpText: 'drink <em>more</em> water'
      }
    );

    // Render it
    document.body.innerHTML = '<div id="placeholder"></div>';
    boundBlock = blockDef.render(
      $('#placeholder'),
      'the-prefix',
      'Test initial state'
    );
  });

  test('it renders correctly', () => {
    expect(document.body.innerHTML).toMatchSnapshot();
  });
});
