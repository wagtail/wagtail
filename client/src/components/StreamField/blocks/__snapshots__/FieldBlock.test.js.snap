// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`telepath: wagtail.blocks.FieldBlock catches widget render errors it renders correctly 1`] = `
"<div class=\\"w-field__wrapper\\" data-field-wrapper=\\"\\">
        <div class=\\"w-field w-field--char_field w-field--text_input w-field--error\\" data-field=\\"\\">
          <div class=\\"w-field__errors\\" id=\\"the-prefix-errors\\" data-field-errors=\\"\\">
            <svg class=\\"icon icon-warning w-field__errors-icon\\" aria-hidden=\\"true\\"><use href=\\"#icon-warning\\"></use></svg>
          <p class=\\"error-message\\"><span>This widget failed to render, please check the console for details</span></p></div>
          <div id=\\"the-prefix-helptext\\" data-field-help=\\"\\"></div>
          <div class=\\"w-field__input\\" data-field-input=\\"\\">
            <div data-streamfield-widget=\\"\\"></div>
          </div>
        </div>
      </div>"
`;

exports[`telepath: wagtail.blocks.FieldBlock it renders correctly 1`] = `
"<div class=\\"w-field__wrapper\\" data-field-wrapper=\\"\\">
        <div class=\\"w-field w-field--char_field w-field--text_input\\" data-field=\\"\\">
          <div class=\\"w-field__errors\\" id=\\"the-prefix-errors\\" data-field-errors=\\"\\">
            <svg class=\\"icon icon-warning w-field__errors-icon\\" aria-hidden=\\"true\\" hidden=\\"\\"><use href=\\"#icon-warning\\"></use></svg>
          </div>
          <div id=\\"the-prefix-helptext\\" data-field-help=\\"\\"><p class=\\"help\\">drink <em>more</em> water</p></div>
          <div class=\\"w-field__input\\" data-field-input=\\"\\">
            <p name=\\"the-prefix\\" id=\\"the-prefix\\">The widget</p>
          </div>
        </div>
      </div>"
`;

exports[`telepath: wagtail.blocks.FieldBlock setError() renders errors 1`] = `
"<div class=\\"w-field__wrapper\\" data-field-wrapper=\\"\\">
        <div class=\\"w-field w-field--char_field w-field--text_input w-field--error\\" data-field=\\"\\">
          <div class=\\"w-field__errors\\" id=\\"the-prefix-errors\\" data-field-errors=\\"\\">
            <svg class=\\"icon icon-warning w-field__errors-icon\\" aria-hidden=\\"true\\"><use href=\\"#icon-warning\\"></use></svg>
          <p class=\\"error-message\\"><span>Field must not contain the letter E</span><span>Field must contain a story about kittens</span></p></div>
          <div id=\\"the-prefix-helptext\\" data-field-help=\\"\\"><p class=\\"help\\">drink <em>more</em> water</p></div>
          <div class=\\"w-field__input\\" data-field-input=\\"\\">
            <p name=\\"the-prefix\\" id=\\"the-prefix\\">The widget</p>
          </div>
        </div>
      </div>"
`;

exports[`telepath: wagtail.blocks.FieldBlock with comments enabled it renders correctly 1`] = `
"<div class=\\"w-field__wrapper\\" data-field-wrapper=\\"\\">
        <div class=\\"w-field w-field--char_field w-field--text_input w-field--commentable\\" data-field=\\"\\">
          <div class=\\"w-field__errors\\" id=\\"the-prefix-errors\\" data-field-errors=\\"\\">
            <svg class=\\"icon icon-warning w-field__errors-icon\\" aria-hidden=\\"true\\" hidden=\\"\\"><use href=\\"#icon-warning\\"></use></svg>
          </div>
          <div id=\\"the-prefix-helptext\\" data-field-help=\\"\\"><p class=\\"help\\">drink <em>more</em> water</p></div>
          <div class=\\"w-field__input\\" data-field-input=\\"\\">
            <p name=\\"the-prefix\\" id=\\"the-prefix\\">The widget</p>
          <button type=\\"button\\" aria-label=\\"Add Comment\\" data-comment-add=\\"\\" class=\\"w-field__comment-button w-field__comment-button--add u-hidden\\"><svg class=\\"icon icon-comment-add \\" aria-hidden=\\"true\\"><use href=\\"#icon-comment-add\\"></use></svg><svg class=\\"icon icon-comment-add-reversed \\" aria-hidden=\\"true\\"><use href=\\"#icon-comment-add-reversed\\"></use></svg></button></div>
        </div>
      </div>"
`;
